### PR TITLE
First pass at document sidebar

### LINKED ISSUE #3784: Sidebar / Document Inspector functionality.
## Feature description

This ticket was made out of this [conversation in slack](https://xwp.slack.com/archives/CCCAUAH6F/p1574088174001600). 

"Old" Editor:
image

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The post visibility: Private functionality is ported into the new editor
* The post visibility: Public functionality is ported into the new editor
* The post visibility: Password Protected functionality is ported into the new editor
* The post status: Published functionality is ported into the new editor
* The post status: Draft functionality is ported into the new editor
* The post status: Scheduled functionality ( selection of publish date ) is ported into the new editor
* The post author functionality is ported into the new editor
* The move post to trash functionality is ported into the new editor
* The post permalink functionality is ported into the new editor
* The post excerpt functionality is ported into the new editor
* The featured image functionality for setting meta tags as primary image is ported into the new editor
* Author assignment and publish fields should only show to user that have this capability. 

## Implementation brief

* Load a list of users into state from REST API
* Load a list of post into state from REST API
* Do capability checks for publish / select author. If user does not have cap, do not show option. 
* Refactor uploadMedia to be more reusable. 
* Add field to REST API to get feature image. 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Fixes #3784

## Checklist

- [x] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: assets/src/edit-story/components/header/buttons.js [hunk 1]
```diff
@@ -71,7 +71,20 @@ function Publish() {
...
-	const text = ( status !== 'publish' ) ? __( 'Publish' ) : __( 'Update' );
+	let text;
+
+	switch ( status ) {
+		case 'publish':
+		case 'private':
+			text = __( 'Update' );
+			break;
+		case 'future':
+			text = __( 'Sceduled' );
```
Comment (swissspidy) [inline assets/src/edit-story/components/header/buttons.js]:
```suggestion
			text = __( 'Scheduled' );
```

File: assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js [hunk 2]
```diff
@@ -12,12 +12,13 @@
...
  * @return {Object} New state
  */
-function restore( state, { pages, current, selection, story } ) {
+function restore( state, { pages, current, selection, story, capabilities } ) {
 	if ( ! Array.isArray( pages ) || pages.length === 0 ) {
 		return state;
 	}
 
 	const newStory = typeof story === 'object' ? story : {};
+	const newCapabilities = typeof story === 'object' ? capabilities : {};
```
Comment (swissspidy) [inline assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js]: Should this be like this?

```suggestion
	const newCapabilities = typeof capabilities === 'object' ? capabilities : {};
```

File: assets/src/edit-story/app/story/effects/useLoadStory.js [hunk 3]
```diff
@@ -25,31 +35,50 @@ function useLoadStory( {
 					status,
 					author,
 					slug,
+					date,
+					modified,
+					excerpt: { raw: excerpt },
```
Comment (miina) [inline assets/src/edit-story/app/story/effects/useLoadStory.js]: I wonder if excerpt is going to be used for Stories? It wasn't in use before at least, IIRC. Perhaps would be good to confirm? (Unless you did already)

File: assets/src/edit-story/app/story/effects/useLoadStory.js [hunk 4]
```diff
@@ -9,6 +14,10 @@ import { useEffect } from '@wordpress/element';
 import { useAPI, useHistory } from '../../';
 import { createPage } from '../../../elements';
 
+const getPerm = ( post, field ) => {
```
Comment (miina) [inline assets/src/edit-story/app/story/effects/useLoadStory.js]: Could we write the name of this func out for clarity or add a comment?

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 5]
```diff
@@ -1,9 +1,187 @@
...
+		state: { meta: { isSaving }, story: { author, status, slug, date, excerpt, featuredMediaUrl, password }, capabilities },
+		actions: { updateStory, deleteStory },
+	} = useStory();
+
+	const { postThumbnails } = useConfig();
+
+	useEffect( () => {
+		loadStatuses();
+		loadUsers();
+	} );
```
Comment (miina) [inline assets/src/edit-story/components/inspector/documentInspector.js]: Should this have dependencies? Or is it supposed to always run?

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 6]
```diff
@@ -1,9 +1,187 @@
...
+
+	useEffect( () => {
+		const states = ( status === 'future' ) ? [ 'pending' ] : [ 'future', 'pending' ];
+		setDisabledStatuses( states );
+	}, [ status, setDisabledStatuses ] );
+
+	const handleChangeStatus = useCallback(
+		( evt ) => 	updateStory( { properties: { status: evt.target.value } } ),
```
Comment (miina) [inline assets/src/edit-story/components/inspector/documentInspector.js]: ^ Indent issue

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 7]
```diff
@@ -1,9 +1,187 @@
...
+
 	return (
-		<h2>
-			{ 'Document' }
-		</h2>
+		<>
+			<h2>
+				{ 'Document' }
+			</h2>
+			{ capabilities && capabilities.hasPublishAction && statuses && <SelectMenu
```
Comment (miina) [inline assets/src/edit-story/components/inspector/documentInspector.js]: Would be great to make the naming a bit more clear, e.g. use `canPublish` instead of `hasPublishAction` etc. Thoughts?

File: assets/src/edit-story/app/config/context.js [hunk 8]
```diff
@@ -3,4 +3,4 @@
  */
 import { createContext } from '@wordpress/element';
 
-export default createContext( { api: {} } );
+export default createContext( { api: {}, statuses: {} } );
```
Comment (miina) [inline assets/src/edit-story/app/config/context.js]: Isn't `statuses` part of `api`?

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 9]
```diff
@@ -1,9 +1,187 @@
...
+				label="Status"
+				name="status"
+				options={ statuses }
+				disabled={ isSaving }
+				disabledOptions={ disabledStatuses }
```
Comment (miina) [inline assets/src/edit-story/components/inspector/documentInspector.js]: Would it be more clear to filter the statuses right away instead of passing a separate `disabledOptions`?

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 10]
```diff
@@ -1,9 +1,187 @@
...
+				value={ status }
+				onChange={ handleChangeStatus }
+			/> }
+			{ capabilities && capabilities.hasPublishAction && status !== 'private' && <InputGroup
```
Comment (miina) [inline assets/src/edit-story/components/inspector/documentInspector.js]: When testing, I currently can't see the option for setting the visibility to Password Protected (just the password field). Not sure if I'm missing something?

File: includes/class-amp-story-post-type.php [hunk 11]
```diff
@@ -407,10 +412,12 @@ public static function load_admin_fonts( $post ) {
 		if ( $post_story_data ) {
 			foreach ( $post_story_data as $page ) {
 				foreach ( $page['elements'] as $element ) {
-					$font = AMP_Fonts::get_font( $element['fontFamily'] );
-					if ( $font && ! in_array( $font['slug'], $font_slugs, true ) ) {
-						$fonts[]      = $font;
-						$font_slugs[] = $font['slug'];
+					if ( isset( $element['fontFamily'] ) ) {
```
Comment (miina) [inline includes/class-amp-story-post-type.php]: Could we do `continue` if `! isset( $element['fontFamily'] )` instead? To reduce nesting and improve readability.

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 12]
```diff
@@ -1,9 +1,187 @@
...
+
+			<RemoveButton onClick={ handleRemoveStory } dangerouslySetInnerHTML={ { __html: 'Move to trash' } } />
+			<InputGroup
+				label={ 'Published date' }
+				type={ 'datetime-local' }
+				value={ date }
+				disabled={ isSaving }
+				onChange={ handleChangeDate }
+			/>
+			{ capabilities && capabilities.hasAssignAuthorAction && users && <SelectMenu
```
Comment (swissspidy) [inline assets/src/edit-story/components/inspector/documentInspector.js]: It seems like something weird is going on with the capability checks.

1. Edit story.
1. Assign story to user with subscriber role
1. Update story
1. Click Undo button.
1. See that the author field and others disappear unless the page is reloaded.

This might be an issue with Undo and not the capability checks added here, but I wanted to flag it either way.

Comment (swissspidy) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Ah, looks like @miina flagged the same thing too :-)

Resolving this as a duplicate.

File: assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js [hunk 13]
```diff
@@ -26,6 +27,7 @@ function restore( state, { pages, current, selection, story } ) {
 		current: newCurrent,
 		selection: newSelection,
 		story: newStory,
+		capabilities: newCapabilities,
```
Comment (barklund) [inline assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js]: Please remember to add it to [the initial state](https://github.com/ampproject/amp-wp/blob/feature/document-sidebar1/assets/src/edit-story/app/story/useStoryReducer/useStoryReducer.js#L12-L17) of the reducer.

Also, please remember to update all tests (ensure 100% test coverage in [`reducers`](https://github.com/ampproject/amp-wp/tree/feature/document-sidebar1/assets/src/edit-story/app/story/useStoryReducer/reducers) folder).

File: assets/src/edit-story/components/header/title.js [hunk 14]
```diff
@@ -25,26 +27,42 @@ const Input = styled.input`
...
+
 	const handleChange = useCallback(
 		( evt ) => updateStory( { properties: { title: evt.target.value } } ),
 		[ updateStory ],
 	);
 
+	const handleBlur = useCallback(
+		() => {
+			if ( ! slug || slug === storyId ) {
+				const clearSlug = encodeURIComponent( cleanForSlug( titleFormatted( title ) ) ) || storyId;
```
Comment (barklund) [inline assets/src/edit-story/components/header/title.js]: I don't understand the name `clearSlug` here.

File: assets/src/edit-story/components/header/title.js [hunk 15]
```diff
@@ -25,26 +27,42 @@ const Input = styled.input`
...
 		return null;
 	}
 
-	// TODO Make sure that Auto Draft checks translations.
-	const titleFormatted = ( status === 'auto-draft' && title === 'Auto Draft' ) ? '' : title;
-
 	return (
 		<Input
-			value={ titleFormatted }
+			value={ titleFormatted( title ) }
```
Comment (barklund) [inline assets/src/edit-story/components/header/title.js]: Doesn't this give warnings about switching from controlled to uncontrolled or is that not an issue?

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 16]
```diff
@@ -1,9 +1,187 @@
...
+
+	const handleChangeExcerpt = useCallback(
+		( value, evt ) => updateStory( { properties: { excerpt: evt.target.value } } ),
+		[ updateStory ],
+	);
+
+	const handleChangeSlug = useCallback(
+		( value, evt ) => updateStory( { properties: { slug: evt.target.value } } ),
+		[ updateStory ],
+	);
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/documentInspector.js]: These can all be generalised to:

```js
	const handleChangeValue = useCallback(
		( prop ) => ( value, evt ) => updateStory( { properties: { [ prop ]: evt.target.value } } ),
		[ updateStory ],
	);
```

And then further down for the individual input:

```js
		onChange={ handleChangeValue( 'slug' ) }
```

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 17]
```diff
@@ -1,9 +1,187 @@
...
+	display: block;
+	align-items: center;
+	margin: 15px 0px;
+`;
+
+const RemoveButton = styled.button`
+	${ ButtonCSS }
+`;
 
 function DocumentInspector() {
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/documentInspector.js]: This component is way too big and should be split up somehow. I don't know what would be a natural split off the top of my head, but it should have concerns split in several smaller components.

File: assets/src/edit-story/components/inspector/inspectorProvider.js [hunk 18]
```diff
@@ -6,26 +6,78 @@ import PropTypes from 'prop-types';
...
 const PREPUBLISH = 'prepublish';
 
 function InspectorProvider( { children } ) {
+	const { actions: { getAllStatuses, getAllUsers } } = useAPI();
 	const [ tab, setTab ] = useState( DESIGN );
+	const [ users, setUsers ] = useState( [] );
+	const [ statuses, setStatuses ] = useState( [] );
+
+	const [ isUsersLoading, setIsUsersLoading ] = useState( false );
+	const [ isStatusesLoading, setIsStatusesLoading ] = useState( false );
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/inspectorProvider.js]: This ought to be changed to two instances of the same reducer-based hook, that can update a list and remember loading state (and probably handle error too).

File: assets/src/edit-story/components/inspector/inspectorProvider.js [hunk 19]
```diff
@@ -6,26 +6,78 @@ import PropTypes from 'prop-types';
...
+	const [ disabledStatuses, setDisabledStatuses ] = useState( [] );
+
+	const loadStatuses = useCallback( () => {
+		if ( ! isStatusesLoading && statuses.length === 0 ) {
+			setIsStatusesLoading( true );
+			getAllStatuses().then( ( data ) => {
+				data = Object.values( data );
+				// eslint-disable-next-line camelcase
+				data = data.filter( ( { show_in_list } ) => show_in_list );
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/inspectorProvider.js]: You could also just rename it inline to avoid the disable such as:

```js
				data = data.filter( ( { show_in_list:isShown } ) => isShown );
```

But yes, it's a bit silly.

File: assets/src/edit-story/components/inspector/inspectorProvider.js [hunk 20]
```diff
@@ -6,26 +6,78 @@ import PropTypes from 'prop-types';
...
 	const [ tab, setTab ] = useState( DESIGN );
+	const [ users, setUsers ] = useState( [] );
+	const [ statuses, setStatuses ] = useState( [] );
+
+	const [ isUsersLoading, setIsUsersLoading ] = useState( false );
+	const [ isStatusesLoading, setIsStatusesLoading ] = useState( false );
+	const [ disabledStatuses, setDisabledStatuses ] = useState( [] );
+
+	const loadStatuses = useCallback( () => {
+		if ( ! isStatusesLoading && statuses.length === 0 ) {
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/inspectorProvider.js]: I'd much rather not call this callback if there already are statuses rather than check every time.

File: assets/src/edit-story/components/inspector/inspectorProvider.js [hunk 21]
```diff
@@ -6,26 +6,78 @@ import PropTypes from 'prop-types';
...
+				} ) => ( {
+					value: id,
+					name,
+				} ) );
+
+				setUsers( saveData );
+				setIsUsersLoading( false );
+			} );
+		}
+	}, [ isUsersLoading, users.length, getAllUsers ] );
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/inspectorProvider.js]: These two callbacks are also very similar now and could be baked into the above-mentioned reducer-based hooks, that will manage their loading. Will make this much cleaner.

File: assets/src/edit-story/components/inspector/documentInspector.js [hunk 22]
```diff
@@ -1,9 +1,187 @@
...
+
+	useEffect( () => {
+		loadStatuses();
+		loadUsers();
+	} );
+
+	useEffect( () => {
+		const states = ( status === 'future' ) ? [ 'pending' ] : [ 'future', 'pending' ];
+		setDisabledStatuses( states );
+	}, [ status, setDisabledStatuses ] );
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/documentInspector.js]: Disabled statuses are only ever used in this component, are they not? Can this just be converted to:

```js
const disabledStatuses = useMemo( () => ( status === 'future' ) ? [ 'pending' ] : [ 'future', 'pending' ], [ status ] );
```

And then everything about disabled statues in the context and provider can be completely deleted? Unless it's required elsewhere of course.

Memoizing isn't even that important, but it looks nice and saves on some re-renders.

**Comment by barklund** [comment]
Really great effort here, adding a lot of new valuable functionality to the editor! Don't let all the comments confuse you, excellent job! 👍

File: assets/src/edit-story/app/config/context.js (hunk 8 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/config/context.js]: Original post statuses came in via config. I didnt remove this. Removed in https://github.com/ampproject/amp-wp/pull/4122/commits/5f9eb6f5b5dc81a15cc5e571b1f3e6412cd9f2a0

File: assets/src/edit-story/app/story/effects/useLoadStory.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/effects/useLoadStory.js]: Added comment here https://github.com/ampproject/amp-wp/pull/4122/commits/6e1e71e9397d71efad6685a01ef9111648c0d876

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 16 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Good catch. Fixed in https://github.com/ampproject/amp-wp/pull/4122/commits/9c4c650652a0962d34eb04c0b0f682773f412aa8

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 7 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: This one was copied from gutenberg - https://github.com/WordPress/gutenberg/blob/94f54892c214ebd014ec126a9575b7e902198ad9/packages/editor/src/components/post-pending-status/check.js#L24 
Not against changing it...

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 9 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Can you explain this, not sure what you mean here.

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 10 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: You are not. Current there is not check to make password protected. Currently, a post becomes password protected by adding a password. I will change this when UX comes in.

File: assets/src/edit-story/components/inspector/inspectorProvider.js (hunk 19 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/inspectorProvider.js]: Works for me. Fixed in https://github.com/ampproject/amp-wp/pull/4122/commits/2ca09099dd8f3274206534faa0fec332ec7890c9

File: includes/class-amp-story-post-type.php (hunk 11 above)
Comment (spacedmonkey) [reply includes/class-amp-story-post-type.php]: Fixed in https://github.com/ampproject/amp-wp/pull/4122/commits/b7aa6d26c7839c7134782ecd0113fd9a66333697

File: assets/src/edit-story/components/inspector/inspectorProvider.js (hunk 18 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/inspectorProvider.js]: Did you have something in mind here?

File: assets/src/edit-story/app/story/effects/useLoadStory.js (hunk 3 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/effects/useLoadStory.js]: @pbakaus Thoughts on excerpt?

File: assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js (hunk 13 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js]: Fixed in https://github.com/ampproject/amp-wp/pull/4122/commits/bae3059721739ba3ffa181c0fda173da440ddff3

File: assets/src/edit-story/components/header/title.js (hunk 14 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/header/title.js]: Changed in https://github.com/ampproject/amp-wp/pull/4122/commits/27a750e4e53b9418c7ea0f6a7adbfdd1806c60d7

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 22 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Good shout. https://github.com/ampproject/amp-wp/pull/4122/commits/f6fd82f3eae64b9f1e3f01f29358ab967618091a

File: assets/src/edit-story/components/inspector/inspectorProvider.js (hunk 18 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/inspectorProvider.js]: This two functions are similar but not the same. The output of the two different api calls are different so require some different logic inside them, making it not easy to refactor into one generate function with params. If I am missing something, please go ahead @barklund give me a code snippet.

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 5 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: These should always run.

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 17 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: I think you are correct here. There is a lot and is makes it a little hard to follow. But until we know the final UX, it seems unwise to split this now. This PR is just so we can get the functionality in, we can style it up and move it around later.

File: assets/src/edit-story/components/header/title.js (hunk 15 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/header/title.js]: Can you explain this a little. I haven't seen any issue as of yet.

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 6 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: This callback has been removed.

**Comment by spacedmonkey** [comment]
> * Undo and Redo don't seem to work for at least some of the fields (didn't test all), e.g. Status and some others.
> * Undo and Redo _do something_ for some of the fields but incorrectly. E.g. changing the slug and then clicking Undo a few times will result in not displaying some of the fields at all:

Good spot @miina and @swissspidy . I have fixed the issue with undo / redo. It was simply a matter of passing the capabilities to the restored state. I had missed a couple of places. See this commit https://github.com/ampproject/amp-wp/pull/4122/commits/62ac276fac3a40c6f2c60b7af27030ccb849d60c

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 12 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Fixed in https://github.com/ampproject/amp-wp/commit/62ac276fac3a40c6f2c60b7af27030ccb849d60c

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 9 above)
Comment (miina) [reply assets/src/edit-story/components/inspector/documentInspector.js]: What I meant is that perhaps instead of having a separate `disabledOptions` attribute, we could directly just filter `statuses` and pass the correct `options` right away, instead of the `SelectMenu` having to be aware of irrelevant options.

E.g. `options={ statuses.filter( status ) => ! disabledStatuses.includes( status ) }` or sth similar.

 Thoughts?

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 5 above)
Comment (miina) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Hmm, perhaps it shouldn't try to load these if there are already statuses/users loaded? Otherwise, it will try to do this with every render.

File: assets/src/edit-story/app/story/effects/useHistoryEntry.js [hunk 23]
```diff
@@ -22,8 +23,9 @@ function useHistoryEntry( {
 			current,
 			pages,
 			selection,
+			capabilities,
 		} );
-	}, [ appendToHistory, story, current, pages, selection ] );
+	}, [ appendToHistory, story, current, pages, selection, capabilities ] );
```
Comment (miina) [inline assets/src/edit-story/app/story/effects/useHistoryEntry.js]: Actually, (I might be missing something but) why are capabilities relevant to history entry? Aren't capabilities static and coming from DB, and "Undo" and "Redo" shouldn't influence the capabilities? Thoughts?

Comment (spacedmonkey) [reply assets/src/edit-story/app/story/effects/useHistoryEntry.js]: ATM capabilities are stored in the story state. If capabilities are not copied when state is changed, then the capabilities lost and the error you saw happens. 

That is an argument to be said that they shouldn't be stored there in the first place. That is another discussion.

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 5 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: No it will not. In those callbacks it checks to see if they have been loaded. See https://github.com/ampproject/amp-wp/pull/4122/files#diff-89184968733e534046330610da4d7328R31

File: assets/src/edit-story/components/inspector/documentInspector.js (hunk 9 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Works for me.

Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/documentInspector.js]: Fixed in https://github.com/ampproject/amp-wp/pull/4122/commits/c47fcf1e39e549cd797892635c1986946bedc0db

File: assets/src/edit-story/app/story/effects/useHistoryEntry.js (hunk 23 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/effects/useHistoryEntry.js]: Editor-wide state was a topic the other day already. Thoughts, @barklund?

Comment (barklund) [reply assets/src/edit-story/app/story/effects/useHistoryEntry.js]: I did look at this one, and didn't really know how to feel about it. On one hand, it doesn't feel like something that belongs inside the story reducer, as it's not in any way part of the story nor is it anything you can edit. But one the other hand, where else would it go?

I would probably prefer this being taken back out if the story reducer, but not completely sure about where to put it instead. Maybe it's an API level thing and belongs in the API state? It's a tricky one.

File: assets/src/edit-story/app/story/effects/useLoadStory.js (hunk 3 above)
Comment (pbakaus) [reply assets/src/edit-story/app/story/effects/useLoadStory.js]: I think this might be useful to support as the excerpt is often used as description of something. But I don't think it's P0.

File: assets/src/edit-story/app/story/effects/useHistoryEntry.js (hunk 23 above)
Comment (miina) [reply assets/src/edit-story/app/story/effects/useHistoryEntry.js]: Sounds good to me to merge as-is for the purposes of this issue, and open a separate ticket for discussing how to store it "properly".

Comment (pbakaus) [reply assets/src/edit-story/app/story/effects/useHistoryEntry.js]: sgtm.

File: assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js [hunk 24]
```diff
@@ -12,12 +12,13 @@
...
  * @return {Object} New state
  */
-function restore( state, { pages, current, selection, story } ) {
+function restore( state, { pages, current, selection, story, capabilities } ) {
 	if ( ! Array.isArray( pages ) || pages.length === 0 ) {
 		return state;
 	}
 
 	const newStory = typeof story === 'object' ? story : {};
+	const newCapabilities = typeof capabilities === 'object' ? capabilities : {};
```
Comment (dvoytenko) [inline assets/src/edit-story/app/story/useStoryReducer/reducers/restore.js]: Just accidentally noticed this. Is this still ok given that `typeof null === 'object'`?

File: assets/src/edit-story/components/header/title.js (hunk 15 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/header/title.js]: Looks like title is never `null`? So there's no controlled/uncontrolled issue?

Comment (barklund) [reply assets/src/edit-story/components/header/title.js]: Seems to work fine, my bad.
